include/certifier_framework.h

```diff
@@ -34,87 +34,60 @@ using std::string;
-      int max_num_tc_;
-      int num_tc_;
-      tagged_claim** tc_;
-      int max_num_tkm_;
-      int num_tkm_;
-      channel_key_message** tkm_;
-      int max_num_blobs_;
-      int num_blobs_;
-      tagged_blob_message** tagged_blob_;
 
     public:
-
+      policy_store(int max_ents);
       policy_store();
-      policy_store(const string enc_alg, int max_trusted_services, int max_trusted_signed_claims,
-          int max_storage_infos, int max_claims, int max_keys, int max_blobs);
       ~policy_store();
 
-      bool replace_policy_key(key_message& k);
-      const key_message* get_policy_key();
+      int get_num_entries();
+      bool add_entry(const string& tag, const string& type, const string& value);
+      int find_entry_by_tag(const string& tag);
+      int find_entry_by_tag_and_type(const string& tag, const string& type);
+      const string* tag(int ent);
+      const string* type(int ent);
+      store_entry* get_entry(int ent);
+      void delete_entry(int ent);
+      void get(int ent, string* v);
+      void put(int ent, const string v);
 
-      int get_num_trusted_services();
-      const trusted_service_message* get_trusted_service_info_by_index(int n);
-      int get_trusted_service_index_by_tag(const string tag);
-      bool add_trusted_service(trusted_service_message& to_add);
-      void delete_trusted_service_by_index(int n);
-
-      int get_num_storage_info();
-      const storage_info_message* get_storage_info_by_index(int n);
-      bool add_storage_info(storage_info_message& to_add);
-      int get_storage_info_index_by_tag(const string& tag);
-      void delete_storage_info_by_index(int n);
-
-      int get_num_claims();
-      const claim_message* get_claim_by_index(int n);
-      bool add_claim(const string& tag, const claim_message& to_add);
-      int get_claim_index_by_tag(const string& tag);
-      void delete_claim_by_index(int n);
-
-      int get_num_signed_claims();
-      const signed_claim_message* get_signed_claim_by_index(int n);
-      int get_signed_claim_index_by_tag(const string& tag);
-      bool add_signed_claim(const string& tag, const signed_claim_message& to_add);
-      void delete_signed_claim_by_index(int n);
-
-      bool add_authentication_key(const string& tag, const key_message& k);
-      const key_message* get_authentication_key_by_tag(const string& tag);
-      const key_message* get_authentication_key_by_index(int index);
-      int get_authentication_key_index_by_tag(const string& tag);
-      void delete_authentication_key_by_index(int index);
-
-      bool add_blob(const string& tag, const string& s);
-      const string* get_blob_by_tag(const string& tag);
-      const string* get_blob_by_index(int index);
-      const tagged_blob_message* get_tagged_blob_info_by_index(int n);
-      int get_blob_index_by_tag(const string& tag);
-      void delete_blob_by_index(int index);
-      int get_num_blobs();
+      bool update_or_insert(const string& tag,
```

Comment:
Looking at your current implementation for `update_or_insert()`, I think if you just have `upsert()` as the single  **public** interface and take away `add_entry()`, the uniqueness check problems will likely be solved on their own.

- `add_entry()` still continues to live as a private interface.
- `upsert` will call `find_entry_by_tag_and_type()`, and if found, it will update the value.
- Otherwise, add a new entry

Currently, these APIs do not seem to support concurrency, so above logic will suffice to ensure uniqueness checks on (tag, type).

Reply:
I want a seperate insert